### ISSUE TITLE
raylib on Raspberry Pi 2 B+ ignoring SetTargetFPS(); 

### ISSUE THREAD

**Issue by ernastovich** [raysan5/raylib#1438]
Hello all, 

I'm having a little bit of trouble with the loop frequency of Raylib on the Raspberry Pi 2 B. The program appears to ignore the SetTargetFPS(##); call. I tested it with several different numbers ranging from extreme values 10 to 200, but there was no change in behavior (neither slowing down, no speeding up). I even tried completely commenting out the SetTargetFPS line, and that also had no adverse effect on the program. 

I'm pretty sure this is a problem with the Raspberry Pi system, I developed and tested the Raylib part of the software on my Mac before finally moving it over to the raspberry pi, and on my Mac it works flawlessly. So something about the raspberry pi system is causing this problem, I'm hoping maybe someone here might know some tricks to try out? 

For the record, I did see some similar issues other people had (although not exact), and one of the recommendations was modifying the raylib busy-wait option. I did try that recommendation to see if it would make a difference, but unfortunately it hasn't. 

Thanks in advance!

**Comment by raysan5** [comment]
What are you seeing if you `DrawFPS()`? Maybe is VSync enabled?

**Comment by ernastovich** [comment]
Ok so I just tried that, here's what happened - 

I had the SetTargetFPS(60); set to 60 FPS
The DrawFPS(); showed it dropping down from 60 to 10, then back up to 12, and it stayed still at 12.

Could you please elaborate on the VSync part? I'm not familiar with what that is, I googled it but come up empty handed. It sounds like VSync is always enabled in RPI by default (?) I could have misunderstood though. I will keep reading up on it in the mean time. 

Thank you for the response!!

**Comment by raysan5** [comment]
VSync is the vertical synchonization, when enabled, the GPU tries to sync your framerate with the display refresh rate. Not sure if it can be enabled on RPI...

Which raylib examples are you testing? Does it happen with all the tested examples?

**Comment by ernastovich** [comment]
You know, I hadn't actually tried any of the example codes. I just tried them now that you mentioned it, and it actually works fine. The FPS dropped to 50 at minimum. Maybe my code is too inefficient and is causing it to lag.

**Comment by raysan5** [comment]
@ernastovich could you share the code you are testing?

**Comment by ernastovich** [comment]
Absolutely. 

Fair warning, I'm an electrical engineer by profession, and an analog engineer at that, so my programming skills are juvenile to say the least. You're about to see some ugly code with insufficient comments :D 

`#include "raylib.h"

#include <time.h>

#include <math.h>

#include <stdlib.h>

#include <stdio.h>

//#include "bcm2835.h"



void speedtape(void);

float inttofloat(int x);

void smallcir(int x[4]);

double degtorad(int x);

int v[4]; //THE FOUR TIRES SPEEDS

int xr; //SPEED SYNTHESIZER

int theta; //steering angle

int north; //compass angle with respect to North

bool direction = true;

bool firststart = true;

void dWheel(void);

int speed =0;

int pixels;

int pixels10=0;

int velocity;

int o;

int t;





Vector2 fl = { 640, 150 }; //these are the center locations for the 4 tire speeds

Vector2 fr = { 896, 150 };

Vector2 rl = { 640, 450 };

Vector2 rr = { 896, 450 };



Vector2 p1 = { 241, 450};

Vector2 p2 = { 271, 450};

Vector2 p3 = { 256, 406};

Vector2 p4 = { 242, 449};

Vector2 p5 = { 270, 449};

Vector2 p6 = { 256, 407};

Vector2 p7 = { 243, 448};

Vector2 p8 = { 269, 448};

Vector2 p9 = { 256, 408};

Vector2 varvec;

Image zro[36];

Rectangle bar={40,300,40,600};

Rectangle bar1 = {16,300,16,600};

Rectangle baru={40,265,40,600};

Rectangle bar1u = {16,265,16,600};

Rectangle barl={40,335,40,600};

Rectangle bar1l = {18,335,18,600};





int main(void)

{

    // Initialization

    //--------------------------------------------------------------------------------------

    const int screenWidth = 1024;

    const int screenHeight = 600;



    InitWindow(screenWidth, screenHeight, "raylib [core] example - input multitouch");



    SetTargetFPS(60);               // Set our game to run at 60 frames-per-second

    //---------------------------------------------------------------------------------------

    xr =0;

    theta = 0;

    dWheel();

    Font menlo = LoadFont("/Users/ernest/Desktop/raylib/Menlo-Regular.ttf");

    

    DrawTextRec(menlo,"0",bar,36,1,true,RAYWHITE);

    DrawTextRec(menlo,"0",bar1,37,1,true,RAYWHITE);

    

    DrawTextRec(menlo,"1",baru,36,1,true,RAYWHITE);

    DrawTextRec(menlo,"1",bar1u,37,1,true,RAYWHITE);

    

    DrawTextRec(menlo,"9",barl,36,1,true,RAYWHITE);

    DrawTextRec(menlo,"9",bar1l,37,1,true,RAYWHITE);

    

    

   

    // Main game loop

    while (!WindowShouldClose())    // Detect window close button or ESC key

    {

        if(xr==360)

        {

            xr=0;

        }

        v[0]=xr; //these is where V is synthetically assigned a non-zero tire speed x4

        v[1]=xr;

        v[2]=xr;

        v[3]=xr;

        

        Color tester = ColorAlpha(BLUE, 0.7); //transparent color

        Color clear = ColorAlpha(BLACK, 0.01); //transparent color

        

        BeginDrawing();



        ClearBackground(BLACK);

        DrawTriangleLines(p3, p1, p2, RAYWHITE);

        DrawTriangleLines(p6, p4, p5, RAYWHITE);

        DrawTriangleLines(p9, p7, p8, RAYWHITE);

        speedtape();

        DrawRectangle(10, 0,50,285,BLACK);

        DrawRectangleGradientV(10,285,50,33,BLACK,clear);

        DrawRectangleGradientV(10,330,50,26,clear,BLACK);

        DrawRectangle(10, 350,50,300,BLACK);

        

        //DrawRectangleLines(10,290,50,20,RAYWHITE);

        //DrawRectangleLines(10,320,50,20,RAYWHITE);



        

        Vector2 cnt={256,800};

        Vector2 ln2 = {256, 300};

        //Vector2 ln3 = {256+300*cos(degtorad(xr)),800+300*sin(degtorad(xr))}; //this is the vector for 0

        //dWheel(ln3.x, ln3.y);

        //DrawCircleSector( cnt, 256, 90, 180, 360, RAYWHITE);

        

        //Rectangle rer = {250, 300, 450, 450};

        //DrawTextRec(GetFontDefault(), "text", rer, 6, 1, false, RAYWHITE);

       // Font menlo = LoadFont("/Users/ernest/Desktop/raylib/Menlo-Regular.ttf");

        //QUEUE THE 3 CHARACTERS, SHOWN, not shown going up, not shown going down

       // Rectangle bar = {40,ones+xr,40,420}; //every 35 pixels is a new number

        



        Vector2 MeasureTextEx(Font font, const char *text, float fontSize, float spacing);

        //DrawTextRec(menlo, TextFormat(MeasureTextEx(menlo, "0123456897",36,1).x),bar,24,1,true,BLUE);

 

        

        DrawRing(cnt,306,303, 0,360,1024,SKYBLUE);

        Font tnr = LoadFont("/System/Library/Fonts/Supplemental/Times New Roman.ttf");

        Rectangle rer1 = {10, 10, 512, 50};

        DrawTextRec(menlo, "GPS COORDINATES: 36.00.000 x 48.00.0000", rer1, 16, 1, false, GREEN);

        

        //Image tsts = ImageTextEx(menlo,"0",6,1,RED);

        //DrawTextureEx(LoadTextureFromImage(tsts), ln2, 0,1,RAYWHITE);

        //DrawTextRec(GetFontDefault()LoadFont("/System/Library/Fonts/Menlo.ttc"), "text", rer, 12, 1, false, RAYWHITE);

        for(int i=0;i<36;i++)

        {

        //consider trying to implement a structure array buildings[buildings] for cleaner code

            int num = xr + i*10;

            int angle = 90 + i*10 + xr;

        Vector2 ln3 = {256+300*cos(degtorad(num)),800+300*sin(degtorad(num))}; //this is the vector for 0

        DrawTextureEx(LoadTextureFromImage(zro[i]), ln3, angle,0.7,RAYWHITE); // SUCCESFUL IMPLEMENTATION OF ROTATED TEXT



        }

        

            //this is the static ring

            DrawRing(fl, 69,71,0,360,400,RAYWHITE);

            DrawRing(fr, 69,71,0,360,400,RAYWHITE);

            DrawRing(rl, 69,71,0,360,400,RAYWHITE);

            DrawRing(rr, 69,71,0,360,400,RAYWHITE);

        

            //Center Divider Line

            DrawLine(512,0,512,600,BLUE);

        

            //this is the precentage round bar

            DrawRing(fl, 60, 70, 90, 90-v[0], 360, tester);

            DrawRing(fr, 60, 70, 90, 90-v[1], 360, tester);

            DrawRing(rl, 60, 70, 90, 90-v[2], 360, tester);

            DrawRing(rr, 60, 70, 90, 90-v[3], 360, tester);

            smallcir(v);

        

            //POSSIBLY this determines the location of the text so that its always centered

            Vector2 siz = MeasureTextEx(menlo, TextFormat("%03.01f",inttofloat(v[0])), 22, 2 );

            Vector2 siz1 = MeasureTextEx(menlo, TextFormat("%03.01f",inttofloat(v[1])), 22, 2 );

            Vector2 siz2 = MeasureTextEx(menlo, TextFormat("%03.01f",inttofloat(v[2])), 22, 2 );

            Vector2 siz3 = MeasureTextEx(menlo, TextFormat("%03.01f",inttofloat(v[3])), 22, 2 );

        

            Vector2 fiz  = { fl.x-(siz.x)/2, fl.y-(siz.y)/2};

            Vector2 fiz1 = { fr.x-(siz1.x)/2, fr.y-(siz1.y)/2};

            Vector2 fiz2 = { rl.x-(siz2.x)/2, rl.y-(siz2.y)/2};

            Vector2 fiz3 = { rr.x-(siz3.x)/2, rr.y-(siz3.y)/2};

            //The following 4 lines of code display the tire rotation speed in percentage

            //DrawText(TextFormat("%03.01f",inttofloat(v[0])) ,fl.x-(siz.x)/2, fl.y-(siz.y)/2, 22, RAYWHITE);

            //DrawText(TextFormat("%03.01f",inttofloat(v[1])) ,fr.x-(siz1.x)/2, fr.y-(siz1.y)/2, 22, RAYWHITE);

            //DrawText(TextFormat("%03.01f",inttofloat(v[2])) ,rl.x-(siz2.x)/2, rl.y-(siz2.y)/2, 22, RAYWHITE);

            //DrawText(TextFormat("%03.01f",inttofloat(v[3])) ,rr.x-(siz3.x)/2, rr.y-(siz3.y)/2, 22, RAYWHITE);

        

            

            DrawTextEx(menlo,TextFormat("%03.01f",inttofloat(v[0])) ,fiz, 22, 1, RAYWHITE);

            DrawTextEx(menlo,TextFormat("%03.01f",inttofloat(v[1])) ,fiz1, 22, 1, RAYWHITE);

            DrawTextEx(menlo,TextFormat("%03.01f",inttofloat(v[2])) ,fiz2, 22, 1, RAYWHITE);

            DrawTextEx(menlo,TextFormat("%03.01f",inttofloat(v[3])) ,fiz3, 22, 1, RAYWHITE);



        

        EndDrawing();

        xr++;

        /*

        if(speed>98)

        {

            direction = false;

        }

        if(speed<98)

        {

            direction = true;

        }

        if(direction)

        {

            speed++;

        }else speed--;

        */

    }

    xr =0;

    /*while(!WindowShouldClose())

    {

        BeginDrawing();

        ClearBackground(BLACK);

        

        DrawTriangleLines(p3, p1, p2, RAYWHITE);

        DrawTriangleLines(p6, p4, p5, RAYWHITE);

        DrawTriangleLines(p9, p7, p8, RAYWHITE);

        EndDrawing();

        xr++;

        

    }*/

    // De-Initialization

    //--------------------------------------------------------------------------------------

    CloseWindow();        // Close window and OpenGL context

    //--------------------------------------------------------------------------------------



    return 0;

}



void smallcir(int x[4])

{

    //BeginDrawing();

    //ClearBackground(BLACK);

    Vector2 notch = {fl.x+70*cos(degtorad(x[0])), fl.y+70*sin(degtorad(x[0]))};

    Vector2 notch1 = {fl.x+60*cos(degtorad(x[0])), fl.y+60*sin(degtorad(x[0]))};

    DrawLineV(notch, notch1, GREEN);

    

    Vector2 notch2 = {fr.x+70*cos(degtorad(x[1])), fr.y+70*sin(degtorad(x[1]))};

    Vector2 notch3 = {fr.x+60*cos(degtorad(x[1])), fr.y+60*sin(degtorad(x[1]))};

    DrawLineV(notch2, notch3, GREEN);

    

    Vector2 notch4 = {rl.x+70*cos(degtorad(x[2])), rl.y+70*sin(degtorad(x[2]))};

    Vector2 notch5 = {rl.x+60*cos(degtorad(x[2])), rl.y+60*sin(degtorad(x[2]))};

    DrawLineV(notch4, notch5, GREEN);

    

    Vector2 notch6 = {rr.x+70*cos(degtorad(x[3])), rr.y+70*sin(degtorad(x[3]))};

    Vector2 notch7 = {rr.x+60*cos(degtorad(x[3])), rr.y+60*sin(degtorad(x[3]))};

    DrawLineV(notch6, notch7, GREEN);

   // EndDrawing();

}



double degtorad(int x)

{

    double rad = (3.14159*x)/180 ;

    return rad;

}

void dWheel()

{

    //varvec.x=x;

    //varvec.y=y;

    

    //onetw = ImageText("120",18,RAYWHITE);

    Font menlo = LoadFont("/Users/ernest/Desktop/raylib/Menlo-Regular.ttf");

    zro[0] = ImageTextEx(menlo,"0",6,1,RED);

    

    zro[1] = ImageTextEx(menlo,"10",6,1,RAYWHITE);

    

    zro[2] = ImageTextEx(menlo,"20",6,1,RAYWHITE);

    

    zro[3] = ImageTextEx(menlo,"30",6,1,RAYWHITE);



    zro[4] = ImageTextEx(menlo,"40",12,1,RAYWHITE);

    

    zro[5] = ImageTextEx(menlo,"50",12,1,RAYWHITE);

    

    zro[6] = ImageTextEx(menlo,"60",12,1,RAYWHITE);

    

    zro[7] = ImageTextEx(menlo,"70",12,1,RAYWHITE);

    

    zro[8] = ImageTextEx(menlo,"80",12,1,RAYWHITE);

    

    zro[9] = ImageTextEx(menlo,"90",12,1,RAYWHITE);

    

    zro[10] = ImageTextEx(menlo,"100",12,1,RAYWHITE);

    

    zro[11] = ImageTextEx(menlo,"110",12,1,RAYWHITE);

    

    zro[12] = ImageTextEx(menlo,"120",12,1,RAYWHITE);

    

    zro[13] = ImageTextEx(menlo,"130",12,1,RAYWHITE);

    

    zro[14] = ImageTextEx(menlo,"140",12,1,RAYWHITE);

    

    zro[15] = ImageTextEx(menlo,"150",12,1,RAYWHITE);

    

    zro[16] = ImageTextEx(menlo,"160",12,1,RAYWHITE);

    

    zro[17] = ImageTextEx(menlo,"170",12,1,RAYWHITE);

    

    zro[18] = ImageTextEx(menlo,"180",12,1,RAYWHITE);

    

    zro[19] = ImageTextEx(menlo,"190",12,1,RAYWHITE);

    

    zro[20] = ImageTextEx(menlo,"200",12,1,RAYWHITE);

    

    zro[21] = ImageTextEx(menlo,"210",12,1,RAYWHITE);

    

    zro[22] = ImageTextEx(menlo,"220",12,1,RAYWHITE);

    

    zro[23] = ImageTextEx(menlo,"230",12,1,RAYWHITE);

    

    zro[24] = ImageTextEx(menlo,"240",12,1,RAYWHITE);

    

    zro[25] = ImageTextEx(menlo,"250",12,1,RAYWHITE);

    

    zro[26] = ImageTextEx(menlo,"260",12,1,RAYWHITE);

    

    zro[27] = ImageTextEx(menlo,"270",12,1,RAYWHITE);

    

    zro[28] = ImageTextEx(menlo,"280",12,1,RAYWHITE);

    

    zro[29] = ImageTextEx(menlo,"290",12,1,RAYWHITE);

    

    zro[30] = ImageTextEx(menlo,"300",12,1,RAYWHITE);

    

    zro[31] = ImageTextEx(menlo,"310",12,1,RAYWHITE);

    

    zro[32] = ImageTextEx(menlo,"320",12,1,RAYWHITE);

    

    zro[33] = ImageTextEx(menlo,"330",12,1,RAYWHITE);

    

    zro[34] = ImageTextEx(menlo,"340",12,1,RAYWHITE);

    

    zro[35] = ImageTextEx(menlo,"350",12,1,RAYWHITE);

    

    /*

     need 36 iterations of

     */

}

float inttofloat(int x)

{

    float pct = (float)x;

    return pct;

}



void speedtape()

{

    

    if(firststart)

    {

        velocity = 0;

        firststart = false;

    }

    int delta = speed - velocity;

    if(delta>0)

    {

        direction = true;

    }else if(delta<0)

    {

        direction = false;

    }

    if(direction)

    {

    if(delta*34>0)

    {

        pixels = delta*34;

    }

    velocity = speed;

    if(pixels !=0)

    {

        bar.y = bar.y + 1;

        baru.y = baru.y +1;

        barl.y++;

        pixels--;

    }

    if(baru.y > 298)

    {

        o = o+1;

        if(o==9&&pixels>33)

        {

            pixels10=34;

        }

        if(o==10)

        {

            o=0;

        }

        bar.y = 300;

        baru.y = 265;

        barl.y = 335;

    }

    if(pixels10>0)

    {

        bar1.y++;

        bar1u.y++;

        bar1l.y++;

        pixels10--;

    }

    if(bar1u.y>298)

    {

        t++;

        if(t==10)

        {

            t=0;

        }

        bar1.y = 300;

        bar1u.y = 265;

        bar1l.y = 335;

    }

    }

    

    // the following code reverses direction

    

    if(!direction)

    {

    if(abs(delta*34)>0)

    {

        pixels = abs(delta*34);

    }

    velocity = speed;

    if(pixels !=0)

    {

        bar.y--;

        baru.y--;

        barl.y--;

        pixels--;

    }

    if(barl.y < 302)

    {

        o--;

        if(o==0&&pixels>33)

        {

            pixels10=34;

        }

        if(o==-1)

        {

            o=9;

        }

        bar.y = 300;

        baru.y = 265;

        barl.y = 335;

    }

    if(pixels10>0)

    {

        bar1.y--;

        bar1u.y--;

        bar1l.y--;

        pixels10--;

    }

    if(bar1l.y<302)

    {

        t--;

        if(t==-1)

        {

            t=9;

        }

        bar1.y = 300;

        bar1u.y = 265;

        bar1l.y = 335;

    }

    }



    const char* ones[10];

    ones[0]="0";

    ones[1]="1";

    ones[2]="2";

    ones[3]="3";

    ones[4]="4";

    ones[5]="5";

    ones[6]="6";

    ones[7]="7";

    ones[8]="8";

    ones[9]="9";

    Font menlo = LoadFont("/Users/ernest/Desktop/raylib/Menlo-Regular.ttf");

    

    DrawTextRec(menlo,ones[o],bar,36,1,true,RAYWHITE);

    DrawTextRec(menlo,ones[t],bar1,38,1,true,RAYWHITE);

    

    if(o==9)

    {

    

    DrawTextRec(menlo,ones[0],baru,36,1,true,RAYWHITE);

    }else

    {

        DrawTextRec(menlo,ones[o+1],baru,36,1,true,RAYWHITE);

    }

    if(t==9)

    {

        DrawTextRec(menlo,ones[0],bar1u,38,1,true,RAYWHITE);

    }else

    {

        

        DrawTextRec(menlo,ones[t+1],bar1u,37,1,true,RAYWHITE);

    }

    if(o==0)

    {

        DrawTextRec(menlo,ones[9],barl,36,1,true,RAYWHITE);

    }else{

    DrawTextRec(menlo,ones[o-1],barl,36,1,true,RAYWHITE);

    }

    if(t==0)

    {

    DrawTextRec(menlo,ones[9],bar1l,38,1,true,RAYWHITE);

    }else

    {

        DrawTextRec(menlo,ones[t-1],bar1l,38,1,true,RAYWHITE);

    }

    /*if(o==9)

    {

        o=0;

    }*/

    

    //DrawTextRec(menlo,"9876543210",bar,36,10,true,RAYWHITE);

    

}`

**Comment by raysan5** [comment]
@ernastovich Sorry, I'm afraid I can not be reviewing that much code, you should provide a short sample illustrating the issue. With a long piece of code the issue is probably on user code and not in raylib.

I take a quick look to this code and you are calling `Draw*()` functions outside `BeginDrawing()`/`EndDrawing()`, that's not recommended. Also, you are calling `LoadFont()` inside the loop several times, that is probably the performance hit. Resources should be loaded at initialization.

I'm closing this issue.

**Comment by ernastovich** [comment]
Ray, completely understandable! And thank you for the pointers, I want to emphasize the example codes within raylib actually worked perfectly! So the problem is indeed within my code. 

I went through the process of continually commenting out bits and pieces of the code and checking the resultant FPS, and I did get it to finally run at 60 FPS. 

Thanks again for the support! Have a happy holiday season!
